File: app/src/main/java/com/alexkn/syntact/domain/worker/CreateBucketWorker.java
package com.alexkn.syntact.domain.worker;

import android.content.Context;
import android.util.Log;

import androidx.annotation.NonNull;
import androidx.work.Data;
import androidx.work.Worker;
import androidx.work.WorkerParameters;

import com.alexkn.syntact.app.ApplicationComponentProvider;
import com.alexkn.syntact.app.Property;
import com.alexkn.syntact.data.common.AppDatabase;
import com.alexkn.syntact.data.dao.BucketDao;
import com.alexkn.syntact.restservice.TemplateType;
import com.alexkn.syntact.data.model.Bucket;
import com.alexkn.syntact.restservice.PhrasesRequest;
import com.alexkn.syntact.restservice.SyntactService;
import com.alexkn.syntact.restservice.Template;
import com.alexkn.syntact.restservice.TemplateRequest;

import java.io.IOException;
import java.time.Instant;
import java.util.Locale;
import java.util.stream.Collectors;
import java.util.stream.Stream;

import javax.inject.Inject;

import okhttp3.ResponseBody;
import retrofit2.Response;

public class CreateBucketWorker extends Worker {

    private static final String TAG = CreateBucketWorker.class.getSimpleName();

    private final BucketDao bucketDao;

    @Inject
    Property property;

    @Inject
    SyntactService syntactService;

    public CreateBucketWorker(@NonNull Context context, @NonNull WorkerParameters workerParams) {

        super(context, workerParams);
        ((ApplicationComponentProvider) context).getApplicationComponent().inject(this);

        bucketDao = AppDatabase.Companion.getDatabase(context).bucketDao();
    }

    @NonNull
    @Override
    public Result doWork() {

        Data inputData = getInputData();

        String name = inputData.getString("name");
        Locale language = new Locale(inputData.getString("language"));
        String[] words = inputData.getStringArray("words");

        String token = "Token " + property.get("api-auth-token");

        TemplateRequest templateRequest = new TemplateRequest();
        templateRequest.setLanguage(language);
        templateRequest.setName(name);
        templateRequest.setTemplateType(TemplateType.CUSTOM);
        try {
            Response<Template> templateResponse = syntactService.postTemplate(token, templateRequest).execute();
            if (templateResponse.isSuccessful()) {
                Template template = templateResponse.body();
                Bucket bucket = new Bucket();
                bucket.setPhrasesUrl(template.getPhrasesUrl());
                bucket.setCreatedAt(Instant.now());
                bucket.setId(Long.valueOf(template.getId()));
                bucket.setItemCount(template.getCount());
                bucket.setLanguage(template.getLanguage());
                bucket.setUserLanguage(Locale.getDefault());
                Log.i(TAG, "New bucket created");

                Response<ResponseBody> phraseResponse = syntactService.postPhrases(token, bucket.getId(), Locale.getDefault().getLanguage(),
                        Stream.of(words).map(s -> new PhrasesRequest(s, language)).collect(Collectors.toList())).execute();

                if (phraseResponse.isSuccessful()) {
                    Log.i(TAG, "New phrases created");
                    bucket.setItemCount(words.length);
                    bucketDao.insert(bucket);
                }
            }
        } catch (IOException e) {
            Log.e(TAG, "Error creating Bucket", e);
            return Result.failure();
        }

        return Result.success();
    }
}

File: app/src/main/java/com/alexkn/syntact/domain/worker/CreateBucketWorker.kt
package com.alexkn.syntact.domain.worker

import android.content.Context
import android.util.Log
import androidx.work.Data
import androidx.work.Worker
import androidx.work.WorkerParameters
import com.alexkn.syntact.app.ApplicationComponentProvider
import com.alexkn.syntact.app.Property
import com.alexkn.syntact.app.TAG
import com.alexkn.syntact.data.common.AppDatabase
import com.alexkn.syntact.data.dao.BucketDao
import com.alexkn.syntact.data.model.Bucket
import com.alexkn.syntact.restservice.PhrasesRequest
import com.alexkn.syntact.restservice.SyntactService
import com.alexkn.syntact.restservice.TemplateRequest
import com.alexkn.syntact.restservice.TemplateType
import java.io.IOException
import java.util.*
import javax.inject.Inject

class CreateBucketWorker(context: Context, workerParams: WorkerParameters) : Worker(context, workerParams) {

    @Inject @JvmField
    var property: Property? = null

    @Inject @JvmField
    var syntactService: SyntactService? = null

    @Inject
    lateinit var bucketDao: BucketDao

    init {
        (context as ApplicationComponentProvider).applicationComponent.inject(this)
    }

    override fun doWork(): Result {

        val inputData = inputData

        val name = inputData.getString("name")!!
        val language = Locale(inputData.getString("language"))
        val words = inputData.getStringArray("words")!!

        val token = "Token " + property!!["api-auth-token"]

        val templateRequest = TemplateRequest(
                name = name,
                templateType = TemplateType.CUSTOM,
                language = language
        )
        try {
            val templateResponse = syntactService!!.postTemplate(token, templateRequest).execute()
            if (templateResponse.isSuccessful) {
                val template = templateResponse.body()!!

                Log.i(TAG, "New bucket created")

                val phraseResponse = syntactService!!.postPhrases(token, template.id.toLong(), Locale.getDefault().language,
                        words.map { PhrasesRequest(it, language) }.toList()).execute()

                if (phraseResponse.isSuccessful) {
                    Log.i(TAG, "New phrases created")
                }
            }
        } catch (e: IOException) {
            Log.e(TAG, "Error creating Bucket", e)
            return Result.failure()
        }

        return Result.success()
    }

}
